Refactor VideoPresentationInterfaceIOS

danae404 opened the pull request:
#### 67862e4bd5cec103b4d1b0d60c359c358999e3a1
<pre>
Refactor VideoPresentationInterfaceIOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=269231">https://bugs.webkit.org/show_bug.cgi?id=269231</a>
<a href="https://rdar.apple.com/122826677">rdar://122826677</a>

Reviewed by Andy Estes.

This patch extracts the AVKit-specific code from VideoPresentationInterfaceIOS
and places it into a subclass called VideoPresentationInterfaceAVKit. There
is also now a second subclass called VideoPresentationInterfaceLMK, which
will be used on the vision platform in the future. For now, all IOS platforms
instantiate VideoPresentationInterfaceAVKit.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceLMK.mm: Removed.
* Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.h:
(WebCore::VideoPresentationInterfaceAVKit::playerLayerView const): Deleted.
(WebCore::VideoPresentationInterfaceAVKit::changingStandbyOnly): Deleted.
(): Deleted.
* Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.mm:
(-[WebAVPlayerViewControllerDelegate fullscreenInterface]):
(-[WebAVPlayerViewControllerDelegate setFullscreenInterface:]):
(convertToExitFullScreenReason):
(-[WebAVPlayerViewController initWithFullscreenInterface:]):
(-[WebAVPlayerViewController configurePlayerViewControllerWithFullscreenInterface:]):
(-[WebAVPlayerViewController logChannel]):
(WebCore::VideoPresentationInterfaceAVKit::create):
(WebCore::VideoPresentationInterfaceAVKit::VideoPresentationInterfaceAVKit):
(WebCore::VideoPresentationInterfaceAVKit::~VideoPresentationInterfaceAVKit):
(WebCore::VideoPresentationInterfaceAVKit::playerViewController const):
(WebCore::VideoPresentationInterfaceAVKit::avPlayerViewController const):
(WebCore::VideoPresentationInterfaceAVKit::setupFullscreen):
(WebCore::VideoPresentationInterfaceAVKit::updateRouteSharingPolicy):
(WebCore::VideoPresentationInterfaceAVKit::setPlayerIdentifier):
(WebCore::VideoPresentationInterfaceAVKit::setupPlayerViewController):
(WebCore::VideoPresentationInterfaceAVKit::invalidatePlayerViewController):
(WebCore::VideoPresentationInterfaceAVKit::presentFullscreen):
(WebCore::VideoPresentationInterfaceAVKit::dismissFullscreen):
(WebCore::VideoPresentationInterfaceAVKit::tryToStartPictureInPicture):
(WebCore::VideoPresentationInterfaceAVKit::stopPictureInPicture):
(WebCore::VideoPresentationInterfaceAVKit::isPlayingVideoInEnhancedFullscreen const):
(WebCore::VideoPresentationInterfaceAVKit::setAllowsPictureInPicturePlayback):
(WebCore::VideoPresentationInterfaceAVKit::setShowsPlaybackControls):
(WebCore::VideoPresentationInterfaceAVKit::setContentDimensions):
(WebCore::setSupportsPictureInPicture):
(WebCore::supportsPictureInPicture):
(clearUIColor): Deleted.
(boolString): Deleted.
(VideoPresentationInterfaceAVKit::create): Deleted.
(VideoPresentationInterfaceAVKit::VideoPresentationInterfaceAVKit): Deleted.
(VideoPresentationInterfaceAVKit::~VideoPresentationInterfaceAVKit): Deleted.
(VideoPresentationInterfaceAVKit::avPlayerViewController const): Deleted.
(VideoPresentationInterfaceAVKit::videoDimensionsChanged): Deleted.
(VideoPresentationInterfaceAVKit::setupFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::setVideoPresentationModel): Deleted.
(VideoPresentationInterfaceAVKit::hasVideoChanged): Deleted.
(VideoPresentationInterfaceAVKit::externalPlaybackChanged): Deleted.
(VideoPresentationInterfaceAVKit::pictureInPictureWasStartedWhenEnteringBackground const): Deleted.
(VideoPresentationInterfaceAVKit::exitFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::cleanupFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::setPlayerIdentifier): Deleted.
(VideoPresentationInterfaceAVKit::requestHideAndExitFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::preparedToReturnToInline): Deleted.
(VideoPresentationInterfaceAVKit::mayAutomaticallyShowVideoPictureInPicture const): Deleted.
(VideoPresentationInterfaceAVKit::willStartPictureInPicture): Deleted.
(VideoPresentationInterfaceAVKit::didStartPictureInPicture): Deleted.
(VideoPresentationInterfaceAVKit::failedToStartPictureInPicture): Deleted.
(VideoPresentationInterfaceAVKit::didStopPictureInPicture): Deleted.
(VideoPresentationInterfaceAVKit::prepareForPictureInPictureStopWithCompletionHandler): Deleted.
(VideoPresentationInterfaceAVKit::shouldExitFullscreenWithReason): Deleted.
(VideoPresentationInterfaceAVKit::setInlineRect): Deleted.
(VideoPresentationInterfaceAVKit::doSetup): Deleted.
(VideoPresentationInterfaceAVKit::doEnterFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::doExitFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::exitFullscreenHandler): Deleted.
(VideoPresentationInterfaceAVKit::enterFullscreenHandler): Deleted.
(VideoPresentationInterfaceAVKit::watchdogTimerFired): Deleted.
(VideoPresentationInterfaceAVKit::isPlayingVideoInEnhancedFullscreen const): Deleted.
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
(WebCore::VideoPresentationInterfaceIOS::playerLayerView const):
(WebCore::VideoPresentationInterfaceIOS::changingStandbyOnly):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(WebCore::clearUIColor):
(WebCore::boolString):
(WebCore::fallbackViewController):
(WebCore::VideoPresentationInterfaceIOS::presentingViewController):
(WebCore::VideoPresentationInterfaceIOS::VideoPresentationInterfaceIOS):
(WebCore::VideoPresentationInterfaceIOS::setVideoPresentationModel):
(WebCore::VideoPresentationInterfaceIOS::setupFullscreen):
(WebCore::VideoPresentationInterfaceIOS::setPlayerIdentifier):
(WebCore::VideoPresentationInterfaceIOS::requestHideAndExitFullscreen):
(WebCore::VideoPresentationInterfaceIOS::preparedToReturnToInline):
(WebCore::VideoPresentationInterfaceIOS::doSetup):
(WebCore::VideoPresentationInterfaceIOS::videoDimensionsChanged):
(WebCore::VideoPresentationInterfaceIOS::externalPlaybackChanged):
(WebCore::VideoPresentationInterfaceIOS::setInlineRect):
(WebCore::VideoPresentationInterfaceIOS::enterFullscreen):
(WebCore::VideoPresentationInterfaceIOS::doEnterFullscreen):
(WebCore::VideoPresentationInterfaceIOS::enterFullscreenHandler):
(WebCore::VideoPresentationInterfaceIOS::exitFullscreen):
(WebCore::VideoPresentationInterfaceIOS::doExitFullscreen):
(WebCore::VideoPresentationInterfaceIOS::exitFullscreenHandler):
(WebCore::VideoPresentationInterfaceIOS::cleanupFullscreen):
(WebCore::VideoPresentationInterfaceIOS::willStartPictureInPicture):
(WebCore::VideoPresentationInterfaceIOS::didStartPictureInPicture):
(WebCore::VideoPresentationInterfaceIOS::failedToStartPictureInPicture):
(WebCore::VideoPresentationInterfaceIOS::didStopPictureInPicture):
(WebCore::VideoPresentationInterfaceIOS::prepareForPictureInPictureStopWithCompletionHandler):
(WebCore::VideoPresentationInterfaceIOS::shouldExitFullscreenWithReason):
(WebCore::VideoPresentationInterfaceIOS::watchdogTimerFired):
(fallbackViewController): Deleted.
(VideoPresentationInterfaceIOS::presentingViewController): Deleted.
(VideoPresentationInterfaceIOS::VideoPresentationInterfaceIOS): Deleted.
(VideoPresentationInterfaceIOS::~VideoPresentationInterfaceIOS): Deleted.
(VideoPresentationInterfaceIOS::playerController const): Deleted.
(VideoPresentationInterfaceIOS::applicationDidBecomeActive): Deleted.
(VideoPresentationInterfaceIOS::enterFullscreen): Deleted.
(VideoPresentationInterfaceIOS::exitFullscreenWithoutAnimationToMode): Deleted.
(VideoPresentationInterfaceIOS::invalidate): Deleted.
(VideoPresentationInterfaceIOS::preparedToExitFullscreen): Deleted.
(VideoPresentationInterfaceIOS::prepareForPictureInPictureStop): Deleted.
(VideoPresentationInterfaceIOS::willStopPictureInPicture): Deleted.
(VideoPresentationInterfaceIOS::setHasVideoContentLayer): Deleted.
(VideoPresentationInterfaceIOS::preparedToReturnToStandby): Deleted.
(VideoPresentationInterfaceIOS::finalizeSetup): Deleted.
(VideoPresentationInterfaceIOS::returnToStandby): Deleted.
(VideoPresentationInterfaceIOS::setMode): Deleted.
(VideoPresentationInterfaceIOS::clearMode): Deleted.
(VideoPresentationInterfaceIOS::logIdentifier const): Deleted.
(VideoPresentationInterfaceIOS::loggerPtr const): Deleted.
(VideoPresentationInterfaceIOS::logChannel const): Deleted.
* Source/WebCore/platform/ios/VideoPresentationInterfaceLMK.h: Removed.
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.h: Renamed from Source/WebCore/platform/ios/PlaybackSessionInterfaceLMK.h.
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm: Added.
(-[WKLinearMediaPlayerDelegate initWithInterface:]):
(WebKit::PlaybackSessionInterfaceLMK::create):
(WebKit::PlaybackSessionInterfaceLMK::PlaybackSessionInterfaceLMK):
(WebKit::PlaybackSessionInterfaceLMK::~PlaybackSessionInterfaceLMK):
(WebKit::PlaybackSessionInterfaceLMK::invalidate):
(WebKit::PlaybackSessionInterfaceLMK::playerController const):
(WebKit::PlaybackSessionInterfaceLMK::durationChanged):
(WebKit::PlaybackSessionInterfaceLMK::currentTimeChanged):
(WebKit::PlaybackSessionInterfaceLMK::bufferedTimeChanged):
(WebKit::PlaybackSessionInterfaceLMK::rateChanged):
(WebKit::PlaybackSessionInterfaceLMK::seekableRangesChanged):
(WebKit::PlaybackSessionInterfaceLMK::canPlayFastReverseChanged):
(WebKit::PlaybackSessionInterfaceLMK::audioMediaSelectionOptionsChanged):
(WebKit::PlaybackSessionInterfaceLMK::legibleMediaSelectionOptionsChanged):
(WebKit::PlaybackSessionInterfaceLMK::externalPlaybackChanged):
(WebKit::PlaybackSessionInterfaceLMK::wirelessVideoPlaybackDisabledChanged):
(WebKit::PlaybackSessionInterfaceLMK::mutedChanged):
(WebKit::PlaybackSessionInterfaceLMK::volumeChanged):
(WebKit::PlaybackSessionInterfaceLMK::logClassName const):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h: Added.
(WebKit::VideoPresentationInterfaceLMK::logClassName const):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm: Renamed from Source/WebCore/platform/ios/VideoPresentationInterfaceLMK.mm.
(-[WKPlayableViewControllerDelegate initWithInterface:]):
(WebKit::VideoPresentationInterfaceLMK::~VideoPresentationInterfaceLMK):
(WebKit::VideoPresentationInterfaceLMK::create):
(WebKit::VideoPresentationInterfaceLMK::VideoPresentationInterfaceLMK):
(WebKit::VideoPresentationInterfaceLMK::linearMediaPlayer const):
(WebKit::VideoPresentationInterfaceLMK::setupFullscreen):
(WebKit::VideoPresentationInterfaceLMK::pictureInPictureWasStartedWhenEnteringBackground const):
(WebKit::VideoPresentationInterfaceLMK::hasVideoChanged):
(WebKit::VideoPresentationInterfaceLMK::setPlayerIdentifier):
(WebKit::VideoPresentationInterfaceLMK::mayAutomaticallyShowVideoPictureInPicture const):
(WebKit::VideoPresentationInterfaceLMK::setupPlayerViewController):
(WebKit::VideoPresentationInterfaceLMK::invalidatePlayerViewController):
(WebKit::VideoPresentationInterfaceLMK::presentFullscreen):
(WebKit::VideoPresentationInterfaceLMK::dismissFullscreen):
(WebKit::VideoPresentationInterfaceLMK::isPlayingVideoInEnhancedFullscreen const):
(WebKit::VideoPresentationInterfaceLMK::avPlayerViewController const):
(WebKit::VideoPresentationInterfaceLMK::playerViewController const):
(WebKit::VideoPresentationInterfaceLMK::setContentDimensions):
(WebKit::VideoPresentationInterfaceLMK::setShowsPlaybackControls):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/274626@main">https://commits.webkit.org/274626@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/95d2b8846c9a435b687acf84a7ebc81029afa0d5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39585 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18564 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41941 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42120 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35485 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41891 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21454 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15893 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33053 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15646 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34261 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13574 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13553 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35214 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43397 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35936 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35545 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39334 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14397 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11857 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37596 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16003 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8873 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16051 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15660 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.mm:
```diff
@@ -765,101 +745,54 @@ - (const Logger*)loggerPtr
 void VideoPresentationInterfaceAVKit::setupFullscreen(UIView& videoView, const FloatRect& initialRect, const FloatSize& videoDimensions, UIView* parentView, HTMLMediaElementEnums::VideoFullscreenMode mode, bool allowsPictureInPicturePlayback, bool standby, bool blocksReturnToFullscreenFromPictureInPicture)
 {
     ASSERT(standby || mode != HTMLMediaElementEnums::VideoFullscreenModeNone);
     LOG(Fullscreen, "VideoPresentationInterfaceAVKit::setupFullscreen(%p)", this);
```
I think we can remove these two lines. They're duplicated in the base class.

aestes commented on Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.mm:
```diff
@@ -765,101 +745,54 @@ - (const Logger*)loggerPtr
-    m_blocksReturnToFullscreenFromPictureInPicture = blocksReturnToFullscreenFromPictureInPicture;
-    setInlineRect(initialRect, true);
-    doSetup();
+    VideoPresentationInterfaceIOS::setupFullscreen(videoView, initialRect, videoDimensions, parentView, mode, allowsPictureInPicturePlayback, standby, blocksReturnToFullscreenFromPictureInPicture);
```
We need to call this before calling into the superclass:
```
[playerController() setContentDimensions:videoDimensions];
```

aestes commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
```diff
@@ -64,31 +63,31 @@ class VideoPresentationInterfaceIOS
-    WEBCORE_EXPORT virtual void videoDimensionsChanged(const FloatSize&) = 0;
-    WEBCORE_EXPORT virtual void setupFullscreen(UIView& videoView, const FloatRect& initialRect, const FloatSize& videoDimensions, UIView* parentView, HTMLMediaElementEnums::VideoFullscreenMode, bool allowsPictureInPicturePlayback, bool standby, bool blocksReturnToFullscreenFromPictureInPicture) = 0;
-    WEBCORE_EXPORT virtual void setPlayerIdentifier(std::optional<MediaPlayerIdentifier>) = 0;
+    WEBCORE_EXPORT virtual void videoDimensionsChanged(const FloatSize&);
```
`videoDimensionsChanged` doesn't need to be virtual.

aestes commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
```diff
@@ -126,27 +125,27 @@ class VideoPresentationInterfaceIOS
-#if PLATFORM(WATCHOS) || PLATFORM(APPLETV)
-    UIViewController *presentingViewController();
+#if PLATFORM(WATCHOS) || PLATFORM(APPLETV) || PLATFORM(VISION)
+    WEBCORE_EXPORT UIViewController *presentingViewController();
```
I know why this is here, but I don't think this change is necessary. Nothing in `PLATFORM(VISION)` calls `presentingViewController()` anymore.

aestes commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
```diff
@@ -33,20 +33,44 @@
+static bool ignoreWatchdogForDebugging = false;
+
+#if PLATFORM(WATCHOS) || PLATFORM(APPLETV) || PLATFORM(VISION)
 static UIViewController *fallbackViewController(UIView *view)
```
I know why this is here, but I don't think this change is necessary. Nothing in PLATFORM(VISION) calls fallbackViewController() anymore.

danae404 replied:
This guard surrounds both fallbackViewController() and presentingViewController(). Do we need presentingViewController() on PLATFORM(VISION)?

aestes replied:
> Do we need presentingViewController() on PLATFORM(VISION)?

No

aestes commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,181 @@
+/*
```
I don't think this file was actually moved from WebCore?

danae404 replied:
Git is showing that it was deleted from WebCore and that a new file was created in WebKit. I'm not sure how to make it know that it was just moved- it seems like it worked for PlaybackSessionInterfaceLMK.h though.

aestes commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,181 @@
+Ref<PlaybackSessionInterfaceLMK> PlaybackSessionInterfaceLMK::create(PlaybackSessionModel& model)
+{
+    Ref interface = adoptRef(*new PlaybackSessionInterfaceLMK(model));
+    interface->initialize();
```
Does this work? Is there a `PlaybackSessionInterfaceIOS::initialize()`?

danae404 replied:
I forgot to add `PlaybackSessionInterfaceIOS::initialize()` - now fixed

aestes commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -28,74 +28,73 @@
 
-using namespace WebCore;
+#import "PlaybackSessionInterfaceLMK.h"
+#import <LinearMediaKit/LinearMediaKit.h> // FIXME: need an SPI header
```
Let's avoid importing this header in this PR. I'd like to wait until a "LinearMediaKitSPI.h" is properly defined.

aestes commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -28,74 +28,73 @@
-VideoPresentationInterfaceLMK::VideoPresentationInterfaceLMK(PlaybackSessionInterfaceLMK& playbackSessionInterface)
-    : VideoPresentationInterfaceIOS(playbackSessionInterface)
-{
+@interface WKPlayableViewControllerDelegate : NSObject <LMPlayableViewControllerDelegate>
```
Let's avoid defining this delegate in this PR.

aestes commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -28,74 +28,73 @@
-void VideoPresentationInterfaceLMK::hasVideoChanged(bool)
+VideoPresentationInterfaceLMK::VideoPresentationInterfaceLMK(PlaybackSessionInterfaceIOS& playbackSessionInterface)
+    : VideoPresentationInterfaceIOS { playbackSessionInterface }
+    , m_playerViewControllerDelegate { adoptNS([[WKPlayableViewControllerDelegate alloc] initWithInterface:*this]) }
```
Let's avoid adding `m_playerViewControllerDelegate` in this PR.

aestes commented on Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.h:
```diff
@@ -0,0 +1,81 @@
+
+class PlaybackSessionInterfaceIOS;
+
+class VideoPresentationInterfaceAVKit : public VideoPresentationInterfaceIOS {
```
nit: `class VideoPresentationInterfaceAVKit final : ...`

aestes commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
```diff
@@ -170,81 +159,79 @@ class VideoPresentationInterfaceIOS final
+        NeedsEnterFullScreen = 1 << 0,
+        NeedsExitFullScreen = 1 << 1,
+    };
+    using NextActions = OptionSet<NextAction>;
```
Nit: a lot of the member variables and functions above have moved from their original location for no good reason. would be nice to maintain their positions when possible.

aestes commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,181 @@
+{
+    [m_player setDuration:duration];
+    [m_player setCanTogglePlayback:YES];
+    [m_player setHasAudioContent:YES];
```
I'd exclude the body of this function for now

aestes commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,181 @@
+
+void PlaybackSessionInterfaceLMK::currentTimeChanged(double currentTime, double)
+{
+    [m_player setCurrentTime:currentTime];
```
I'd exclude the body of this function for now

aestes commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,181 @@
+    if (playbackState.contains(PlaybackSessionModel::PlaybackState::Playing))
+        [m_player setPlaybackRate:playbackRate];
+    else
+        [m_player setPlaybackRate:0];
```
I'd exclude the body of this function for now

aestes commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    [m_player setSeekableTimeRanges:seekableRanges.get()];
+    [m_player setCanSeek:[seekableRanges count] != 0];
```
I'd exclude the body of this function for now

aestes commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,181 @@
+
+void PlaybackSessionInterfaceLMK::canPlayFastReverseChanged(bool canPlayFastReverse)
+{
+    [m_player setCanScanBackward:canPlayFastReverse];
```
I'd exclude the body of this function for now

aestes commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,181 @@
+
+void PlaybackSessionInterfaceLMK::mutedChanged(bool muted)
+{
+    [m_player setIsMuted:muted];
```
I'd exclude the body of this function for now

aestes commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,181 @@
+
+void PlaybackSessionInterfaceLMK::volumeChanged(double volume)
+{
+    [m_player setVolume:volume];
```
I'd exclude the body of this function for now

aestes reviewed (approved):
looks great! r=me once comments are addressed.

danae404 committed:
Refactor VideoPresentationInterfaceIOS
https://bugs.webkit.org/show_bug.cgi?id=269231
rdar://122826677

Reviewed by Andy Estes.

This patch extracts the AVKit-specific code from VideoPresentationInterfaceIOS
and places it into a subclass called VideoPresentationInterfaceAVKit. There
is also now a second subclass called VideoPresentationInterfaceLMK, which
will be used on the vision platform in the future. For now, all IOS platforms
instantiate VideoPresentationInterfaceAVKit.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceLMK.mm: Removed.
* Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.h:
(WebCore::VideoPresentationInterfaceAVKit::playerLayerView const): Deleted.
(WebCore::VideoPresentationInterfaceAVKit::changingStandbyOnly): Deleted.
(): Deleted.
* Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.mm:
(-[WebAVPlayerViewControllerDelegate fullscreenInterface]):
(-[WebAVPlayerViewControllerDelegate setFullscreenInterface:]):
(convertToExitFullScreenReason):
(-[WebAVPlayerViewController initWithFullscreenInterface:]):
(-[WebAVPlayerViewController configurePlayerViewControllerWithFullscreenInterface:]):
(-[WebAVPlayerViewController logChannel]):
(WebCore::VideoPresentationInterfaceAVKit::create):
(WebCore::VideoPresentationInterfaceAVKit::VideoPresentationInterfaceAVKit):
(WebCore::VideoPresentationInterfaceAVKit::~VideoPresentationInterfaceAVKit):
(WebCore::VideoPresentationInterfaceAVKit::playerViewController const):
(WebCore::VideoPresentationInterfaceAVKit::avPlayerViewController const):
(WebCore::VideoPresentationInterfaceAVKit::setupFullscreen):
(WebCore::VideoPresentationInterfaceAVKit::updateRouteSharingPolicy):
(WebCore::VideoPresentationInterfaceAVKit::setPlayerIdentifier):
(WebCore::VideoPresentationInterfaceAVKit::setupPlayerViewController):
(WebCore::VideoPresentationInterfaceAVKit::invalidatePlayerViewController):
(WebCore::VideoPresentationInterfaceAVKit::presentFullscreen):
(WebCore::VideoPresentationInterfaceAVKit::dismissFullscreen):
(WebCore::VideoPresentationInterfaceAVKit::tryToStartPictureInPicture):
(WebCore::VideoPresentationInterfaceAVKit::stopPictureInPicture):
(WebCore::VideoPresentationInterfaceAVKit::isPlayingVideoInEnhancedFullscreen const):
(WebCore::VideoPresentationInterfaceAVKit::setAllowsPictureInPicturePlayback):
(WebCore::VideoPresentationInterfaceAVKit::setShowsPlaybackControls):
(WebCore::VideoPresentationInterfaceAVKit::setContentDimensions):
(WebCore::setSupportsPictureInPicture):
(WebCore::supportsPictureInPicture):
(clearUIColor): Deleted.
(boolString): Deleted.
(VideoPresentationInterfaceAVKit::create): Deleted.
(VideoPresentationInterfaceAVKit::VideoPresentationInterfaceAVKit): Deleted.
(VideoPresentationInterfaceAVKit::~VideoPresentationInterfaceAVKit): Deleted.
(VideoPresentationInterfaceAVKit::avPlayerViewController const): Deleted.
(VideoPresentationInterfaceAVKit::videoDimensionsChanged): Deleted.
(VideoPresentationInterfaceAVKit::setupFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::setVideoPresentationModel): Deleted.
(VideoPresentationInterfaceAVKit::hasVideoChanged): Deleted.
(VideoPresentationInterfaceAVKit::externalPlaybackChanged): Deleted.
(VideoPresentationInterfaceAVKit::pictureInPictureWasStartedWhenEnteringBackground const): Deleted.
(VideoPresentationInterfaceAVKit::exitFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::cleanupFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::setPlayerIdentifier): Deleted.
(VideoPresentationInterfaceAVKit::requestHideAndExitFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::preparedToReturnToInline): Deleted.
(VideoPresentationInterfaceAVKit::mayAutomaticallyShowVideoPictureInPicture const): Deleted.
(VideoPresentationInterfaceAVKit::willStartPictureInPicture): Deleted.
(VideoPresentationInterfaceAVKit::didStartPictureInPicture): Deleted.
(VideoPresentationInterfaceAVKit::failedToStartPictureInPicture): Deleted.
(VideoPresentationInterfaceAVKit::didStopPictureInPicture): Deleted.
(VideoPresentationInterfaceAVKit::prepareForPictureInPictureStopWithCompletionHandler): Deleted.
(VideoPresentationInterfaceAVKit::shouldExitFullscreenWithReason): Deleted.
(VideoPresentationInterfaceAVKit::setInlineRect): Deleted.
(VideoPresentationInterfaceAVKit::doSetup): Deleted.
(VideoPresentationInterfaceAVKit::doEnterFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::doExitFullscreen): Deleted.
(VideoPresentationInterfaceAVKit::exitFullscreenHandler): Deleted.
(VideoPresentationInterfaceAVKit::enterFullscreenHandler): Deleted.
(VideoPresentationInterfaceAVKit::watchdogTimerFired): Deleted.
(VideoPresentationInterfaceAVKit::isPlayingVideoInEnhancedFullscreen const): Deleted.
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
(WebCore::VideoPresentationInterfaceIOS::playerLayerView const):
(WebCore::VideoPresentationInterfaceIOS::changingStandbyOnly):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(WebCore::clearUIColor):
(WebCore::boolString):
(WebCore::fallbackViewController):
(WebCore::VideoPresentationInterfaceIOS::presentingViewController):
(WebCore::VideoPresentationInterfaceIOS::VideoPresentationInterfaceIOS):
(WebCore::VideoPresentationInterfaceIOS::setVideoPresentationModel):
(WebCore::VideoPresentationInterfaceIOS::setupFullscreen):
(WebCore::VideoPresentationInterfaceIOS::setPlayerIdentifier):
(WebCore::VideoPresentationInterfaceIOS::requestHideAndExitFullscreen):
(WebCore::VideoPresentationInterfaceIOS::preparedToReturnToInline):
(WebCore::VideoPresentationInterfaceIOS::doSetup):
(WebCore::VideoPresentationInterfaceIOS::videoDimensionsChanged):
(WebCore::VideoPresentationInterfaceIOS::externalPlaybackChanged):
(WebCore::VideoPresentationInterfaceIOS::setInlineRect):
(WebCore::VideoPresentationInterfaceIOS::enterFullscreen):
(WebCore::VideoPresentationInterfaceIOS::doEnterFullscreen):
(WebCore::VideoPresentationInterfaceIOS::enterFullscreenHandler):
(WebCore::VideoPresentationInterfaceIOS::exitFullscreen):
(WebCore::VideoPresentationInterfaceIOS::doExitFullscreen):
(WebCore::VideoPresentationInterfaceIOS::exitFullscreenHandler):
(WebCore::VideoPresentationInterfaceIOS::cleanupFullscreen):
(WebCore::VideoPresentationInterfaceIOS::willStartPictureInPicture):
(WebCore::VideoPresentationInterfaceIOS::didStartPictureInPicture):
(WebCore::VideoPresentationInterfaceIOS::failedToStartPictureInPicture):
(WebCore::VideoPresentationInterfaceIOS::didStopPictureInPicture):
(WebCore::VideoPresentationInterfaceIOS::prepareForPictureInPictureStopWithCompletionHandler):
(WebCore::VideoPresentationInterfaceIOS::shouldExitFullscreenWithReason):
(WebCore::VideoPresentationInterfaceIOS::watchdogTimerFired):
(fallbackViewController): Deleted.
(VideoPresentationInterfaceIOS::presentingViewController): Deleted.
(VideoPresentationInterfaceIOS::VideoPresentationInterfaceIOS): Deleted.
(VideoPresentationInterfaceIOS::~VideoPresentationInterfaceIOS): Deleted.
(VideoPresentationInterfaceIOS::playerController const): Deleted.
(VideoPresentationInterfaceIOS::applicationDidBecomeActive): Deleted.
(VideoPresentationInterfaceIOS::enterFullscreen): Deleted.
(VideoPresentationInterfaceIOS::exitFullscreenWithoutAnimationToMode): Deleted.
(VideoPresentationInterfaceIOS::invalidate): Deleted.
(VideoPresentationInterfaceIOS::preparedToExitFullscreen): Deleted.
(VideoPresentationInterfaceIOS::prepareForPictureInPictureStop): Deleted.
(VideoPresentationInterfaceIOS::willStopPictureInPicture): Deleted.
(VideoPresentationInterfaceIOS::setHasVideoContentLayer): Deleted.
(VideoPresentationInterfaceIOS::preparedToReturnToStandby): Deleted.
(VideoPresentationInterfaceIOS::finalizeSetup): Deleted.
(VideoPresentationInterfaceIOS::returnToStandby): Deleted.
(VideoPresentationInterfaceIOS::setMode): Deleted.
(VideoPresentationInterfaceIOS::clearMode): Deleted.
(VideoPresentationInterfaceIOS::logIdentifier const): Deleted.
(VideoPresentationInterfaceIOS::loggerPtr const): Deleted.
(VideoPresentationInterfaceIOS::logChannel const): Deleted.
* Source/WebCore/platform/ios/VideoPresentationInterfaceLMK.h: Removed.
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.h: Renamed from Source/WebCore/platform/ios/PlaybackSessionInterfaceLMK.h.
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm: Added.
(-[WKLinearMediaPlayerDelegate initWithInterface:]):
(WebKit::PlaybackSessionInterfaceLMK::create):
(WebKit::PlaybackSessionInterfaceLMK::PlaybackSessionInterfaceLMK):
(WebKit::PlaybackSessionInterfaceLMK::~PlaybackSessionInterfaceLMK):
(WebKit::PlaybackSessionInterfaceLMK::invalidate):
(WebKit::PlaybackSessionInterfaceLMK::playerController const):
(WebKit::PlaybackSessionInterfaceLMK::durationChanged):
(WebKit::PlaybackSessionInterfaceLMK::currentTimeChanged):
(WebKit::PlaybackSessionInterfaceLMK::bufferedTimeChanged):
(WebKit::PlaybackSessionInterfaceLMK::rateChanged):
(WebKit::PlaybackSessionInterfaceLMK::seekableRangesChanged):
(WebKit::PlaybackSessionInterfaceLMK::canPlayFastReverseChanged):
(WebKit::PlaybackSessionInterfaceLMK::audioMediaSelectionOptionsChanged):
(WebKit::PlaybackSessionInterfaceLMK::legibleMediaSelectionOptionsChanged):
(WebKit::PlaybackSessionInterfaceLMK::externalPlaybackChanged):
(WebKit::PlaybackSessionInterfaceLMK::wirelessVideoPlaybackDisabledChanged):
(WebKit::PlaybackSessionInterfaceLMK::mutedChanged):
(WebKit::PlaybackSessionInterfaceLMK::volumeChanged):
(WebKit::PlaybackSessionInterfaceLMK::logClassName const):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h: Added.
(WebKit::VideoPresentationInterfaceLMK::logClassName const):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm: Renamed from Source/WebCore/platform/ios/VideoPresentationInterfaceLMK.mm.
(-[WKPlayableViewControllerDelegate initWithInterface:]):
(WebKit::VideoPresentationInterfaceLMK::~VideoPresentationInterfaceLMK):
(WebKit::VideoPresentationInterfaceLMK::create):
(WebKit::VideoPresentationInterfaceLMK::VideoPresentationInterfaceLMK):
(WebKit::VideoPresentationInterfaceLMK::linearMediaPlayer const):
(WebKit::VideoPresentationInterfaceLMK::setupFullscreen):
(WebKit::VideoPresentationInterfaceLMK::pictureInPictureWasStartedWhenEnteringBackground const):
(WebKit::VideoPresentationInterfaceLMK::hasVideoChanged):
(WebKit::VideoPresentationInterfaceLMK::setPlayerIdentifier):
(WebKit::VideoPresentationInterfaceLMK::mayAutomaticallyShowVideoPictureInPicture const):
(WebKit::VideoPresentationInterfaceLMK::setupPlayerViewController):
(WebKit::VideoPresentationInterfaceLMK::invalidatePlayerViewController):
(WebKit::VideoPresentationInterfaceLMK::presentFullscreen):
(WebKit::VideoPresentationInterfaceLMK::dismissFullscreen):
(WebKit::VideoPresentationInterfaceLMK::isPlayingVideoInEnhancedFullscreen const):
(WebKit::VideoPresentationInterfaceLMK::avPlayerViewController const):
(WebKit::VideoPresentationInterfaceLMK::playerViewController const):
(WebKit::VideoPresentationInterfaceLMK::setContentDimensions):
(WebKit::VideoPresentationInterfaceLMK::setShowsPlaybackControls):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/274626@main
